File: app copy/core/agent/polymarket.py
"""
polymarket.py
=============
Fetch predictive market data from Polymarket's Gamma API.
Focuses on Macro, Finance, Geopolitics and Tech/AI — no crypto.
"""
from __future__ import annotations

import json
import urllib.request
from datetime import datetime, timezone
from typing import Optional

_POLYMARKET_GAMMA_API = "https://gamma-api.polymarket.com/events"
_USER_AGENT = "Mozilla/5.0 (compatible; InvestimentosBot/1.0; +https://github.com)"

# ── Keyword lists for category classification ──────────────────────────────

_CRYPTO_KW = [
    "bitcoin", "btc", " eth ", "ethereum", "crypto", "solana", "sol ",
    "binance", "kraken", "coinbase", "dogecoin", "xrp", "ripple",
    "stablecoin", "defi", "nft", "blockchain", "altcoin", "memecoin",
    " bnb ", "polygon", "avalanche", "avax", "litecoin",
]
# Removed false-positive tokens: "matic" (hits "diplomatic","automatic"),
# "eth" alone (hits "method","ethical"), "doge"/"ltc" (covered by full names),
# "bnb" unspaced (hits "airbnb") → replaced with " bnb "

_MACRO_KW = [
    "fed ", "federal reserve", "interest rate", "rate cut", "rate hike",
    "inflation", "recession", "gdp", "unemployment", "cpi", "pce",
    "treasury", "bond yield", "debt ceiling", "deficit", "imf", "world bank",
    "dollar", "usd", "euro ", "yen ", "brl", "ibovespa", "selic",
    "tariff", "trade war", "sanctions", "opec", "oil price", "crude",
    "fomc", "powell", "ecb", "boe ", "boj ", "central bank",
    "s&p 500", "nasdaq", "dow jones", "stock market", "bear market", "bull market",
    "ipo", "earnings", "revenue", "profit", "layoffs",
]

_GEO_KW = [
    "election", "president", "trump", "harris", "biden", "congress",
    "senate", "war", "conflict", "russia", "ukraine", "china ", "taiwan",
    "israel", "iran", "north korea", "nato", "g7", "g20",
    "nuclear", "ceasefire", "peace deal", "coup", "prime minister",
    "vote", "poll", "referendum", "sanction",
]

_TECH_AI_KW = [
    "openai", "gpt", "chatgpt", "agi", "artificial intelligence", "ai model",
    "llm", "claude", "gemini", "mistral", "anthropic",
    "apple", "nvidia", "meta ", "google", "microsoft", "amazon", "tesla",
    "spacex", "robotics", "autonomous", "self-driving", "quantum",
    "regulation ai", "ai regulation", "deepmind", "sam altman",
]

# Category metadata used in UI
CATEGORY_META = {
    "🏦 Macro & Finanças": {
        "color": "#38bdf8",
        "css_class": "macro",
        "desc": "Fed, juros, mercados, câmbio e economia global",
    },
    "🌍 Geopolítica": {
        "color": "#f59e0b",
        "css_class": "geo",
        "desc": "Eleições, conflitos, relações internacionais",
    },
    "🤖 Tech & IA": {
        "color": "#a78bfa",
        "css_class": "tech",
        "desc": "OpenAI, big tech, regulação de IA, inovação",
    },
    "⭐ Em Destaque": {
        "color": "#34d399",
        "css_class": "destaque",
        "desc": "Mercados de alto volume sem categoria definida",
    },
}


def _days_remaining(end_date_str: str) -> Optional[int]:
    """Return days until end_date, or None if unparseable or already past."""
    if not end_date_str:
        return None
    try:
        for fmt in ("%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%dT%H:%M:%S.%fZ", "%Y-%m-%dT%H:%M:%S"):
            try:
                dt = datetime.strptime(end_date_str[:26], fmt)
                break
            except ValueError:
                continue
        else:
            return None
        dt = dt.replace(tzinfo=timezone.utc)
        delta = (dt - datetime.now(timezone.utc)).days
        return None if delta < 0 else delta
    except Exception:
        return None


_BINARY_LABELS = {"yes", "no", "sim", "não", "nao"}


def _parse_json_field(raw) -> list:
    """Parse a field that may be a native list or a JSON-encoded string."""
    if isinstance(raw, list):
        return raw
    try:
        return json.loads(raw) if raw else []
    except (json.JSONDecodeError, TypeError):
        return []


def _label_from_market(m: dict) -> str:
    """
    Best short label for one sub-market inside a categorical event.
    Priority: groupItemTitle → first non-binary outcome name → question (trimmed).
    """
    label = (m.get("groupItemTitle") or "").strip()
    if label:
        return label

    outcomes = _parse_json_field(m.get("outcomes", "[]"))
    for o in outcomes:
        if o.strip().lower() not in _BINARY_LABELS:
            return o.strip()

    q = (m.get("question") or "").strip().rstrip("?")
    for prefix in ("Will ", "Does ", "Is ", "Can ", "Has ", "Did ", "Do "):
        if q.startswith(prefix):
            q = q[len(prefix):]
            break
    return q[:50] if q else ""


def _yes_price(m: dict) -> Optional[float]:
    """
    Return the 'Yes' implied probability (0-1) for a binary sub-market.
    Prefers the outcome explicitly labelled Yes/Sim; falls back to index 0.
    """
    outcomes = _parse_json_field(m.get("outcomes", "[]"))
    prices   = _parse_json_field(m.get("outcomePrices", "[]"))
    if not outcomes or not prices:
        return None

    # Find the index of the "Yes" / "Sim" outcome
    yes_idx = next(
        (i for i, o in enumerate(outcomes) if o.strip().lower() in ("yes", "sim")),
        0,
    )
    try:
        pv = float(prices[yes_idx])
        return pv if 0.0 < pv < 1.0 else None
    except (ValueError, TypeError, IndexError):
        return None


def fetch_polymarket_events(limit: int = 60) -> list[dict]:
    """
    Fetches active, non-closed market prediction events from Polymarket.

    Handles two event shapes from the Gamma API:

    1. **Single-market** (binary or scalar): the market itself carries
       `outcomes` + `outcomePrices` with all answer options — read them directly.

    2. **Multi-market categorical**: the event has N sub-markets, each
       representing ONE possible outcome (e.g. each presidential candidate).
       Every sub-market is internally binary (Yes/No).  We extract the
       "Yes" price from each sub-market and use `groupItemTitle` as the
       outcome label.  This is the case that previously showed only "Yes/No".

    Returns cleaned list sorted by volume descending.
    """
    url = f"{_POLYMARKET_GAMMA_API}?limit={limit}&active=true&order=volume_24hr&ascending=false"
    try:
        req = urllib.request.Request(url, headers={"User-Agent": _USER_AGENT})
        with urllib.request.urlopen(req, timeout=10) as response:
            data = json.loads(response.read().decode("utf-8"))
        if not isinstance(data, list):
            data = data.get("data") or data.get("events") or []
    except Exception as _e:
        import sys
        print(f"[polymarket] fetch failed: {type(_e).__name__}: {_e}", file=sys.stderr)
        return []

    processed: list[dict] = []

    for event in data:
        try:
            markets = event.get("markets") or []
            if not markets:
                continue

            # Total volume across all sub-markets
            total_volume = sum(
                float(m.get("volume") or 0) for m in markets
                if m.get("volume") is not None
            )

            odds: list[dict] = []

            if len(markets) == 1:
                m        = markets[0]
                outcomes = _parse_json_field(m.get("outcomes", "[]"))
                prices   = _parse_json_field(m.get("outcomePrices", "[]"))
                for out, price_str in zip(outcomes, prices):
                    try:
                        pv = float(price_str)
                        if 0.0 < pv < 1.0:
                            odds.append({
                                "outcome": out,
                                "price":   pv,
                                "percent": round(pv * 100, 1),
                            })
                    except (ValueError, TypeError):
                        continue

            else:
                for m in markets:
                    label = _label_from_market(m)
                    if not label:
                        continue
                    pv = _yes_price(m)
                    if pv is None:
                        continue
                    odds.append({
                        "outcome": label,
                        "price":   pv,
                        "percent": round(pv * 100, 1),
                    })

            odds.sort(key=lambda x: x["percent"], reverse=True)
            # Skip if empty or if one outcome has ~100% — market is effectively resolved
            if not odds or odds[0]["percent"] >= 99:
                continue

            end_date_str = event.get("endDate", "")
            days_left    = _days_remaining(end_date_str)

            # Note: active=true in the API already excludes resolved/closed markets.
            # We do NOT filter by days_left here because many high-volume Polymarket
            # events stay in "pending resolution" state with endDate in the past while
            # still being tradeable and interesting to display.

            processed.append({
                "id":          event.get("id"),
                "title":       event.get("title", "Unknown Event"),
                "description": event.get("description") or "",
                "slug":        event.get("slug", ""),
                "url":         f"https://polymarket.com/event/{event.get('slug', '')}",
                "volume":      total_volume,
                "end_date":    end_date_str,
                "days_left":   days_left,
                "odds":        odds,
                "is_binary":   len(odds) == 2,
            })
        except Exception:
            continue

    processed.sort(key=lambda x: x["volume"], reverse=True)
    return processed


def get_curated_crypto_politics_markets() -> dict[str, list[dict]]:
    """
    Fetches and categorises markets for the investment dashboard.
    Excludes crypto. Focuses on Macro/Finance, Geopolitics and Tech/AI.
    """
    events = fetch_polymarket_events(limit=200)

    categories: dict[str, list[dict]] = {
        "🏦 Macro & Finanças": [],
        "🌍 Geopolítica": [],
        "🤖 Tech & IA": [],
        "⭐ Em Destaque": [],
    }

    for ev in events:
        text = (ev["title"] + " " + ev["description"]).lower()

        # Hard-exclude crypto events regardless of other matches
        if any(kw in text for kw in _CRYPTO_KW):
            continue

        if any(kw in text for kw in _MACRO_KW):
            categories["🏦 Macro & Finanças"].append(ev)
        elif any(kw in text for kw in _GEO_KW):
            categories["🌍 Geopolítica"].append(ev)
        elif any(kw in text for kw in _TECH_AI_KW):
            categories["🤖 Tech & IA"].append(ev)
        else:
            categories["⭐ Em Destaque"].append(ev)

    # Top 5 per category sorted by volume (already sorted globally)
    for cat in categories:
        categories[cat] = categories[cat][:10]

    return {k: v for k, v in categories.items() if v}


if __name__ == "__main__":
    markets = get_curated_crypto_politics_markets()
    for cat, evs in markets.items():
        print(f"\n--- {cat} ({len(evs)}) ---")
        for ev in evs:
            days = f"{ev['days_left']}d" if ev["days_left"] is not None else "?"
            print(f"  [{days}] {ev['title']} (Vol: ${ev['volume']:,.0f})")
            for odd in ev["odds"][:2]:
                print(f"       {odd['outcome']}: {odd['percent']}%")

File: lib/polymarket.ts
export interface PolyOdds { outcome: string; price: number; percent: number }
export interface PolyEvent {
  id: string;
  title: string;
  url: string;
  volume: number;
  end_date: string;
  days_left: number | null;
  odds: PolyOdds[];
  is_binary: boolean;
}
export interface PolyResponse {
  categories: Record<string, PolyEvent[]>;
  cached_at: string;
  total_fetched?: number;
  total_parsed?: number;
}

const GAMMA_API = "https://gamma-api.polymarket.com/events";

// ── Keyword classification lists ──────────────────────────────────────────────

const CRYPTO_KW = [
  "bitcoin", "btc", " eth ", "ethereum", "crypto", "solana", "sol ",
  "binance", "kraken", "coinbase", "dogecoin", "xrp", "ripple",
  "stablecoin", "defi", "nft", "blockchain", "altcoin", "memecoin",
  " bnb ", "polygon", "avalanche", "avax", "litecoin",
];

const MACRO_KW = [
  "fed ", "federal reserve", "interest rate", "rate cut", "rate hike",
  "inflation", "recession", "gdp", "unemployment", "cpi", "pce",
  "treasury", "bond yield", "debt ceiling", "deficit", "imf", "world bank",
  "dollar", "usd", "euro ", "yen ", "brl", "ibovespa", "selic",
  "tariff", "trade war", "sanctions", "opec", "oil price", "crude",
  "fomc", "powell", "ecb", "boe ", "boj ", "central bank",
  "s&p 500", "nasdaq", "dow jones", "stock market", "bear market", "bull market",
  "ipo", "earnings", "revenue", "profit", "layoffs",
];

const GEO_KW = [
  "election", "president", "trump", "harris", "biden", "congress",
  "senate", "war", "conflict", "russia", "ukraine", "china ", "taiwan",
  "israel", "iran", "north korea", "nato", "g7", "g20",
  "nuclear", "ceasefire", "peace deal", "coup", "prime minister",
  "vote", "poll", "referendum", "sanction",
];

const TECH_AI_KW = [
  "openai", "gpt", "chatgpt", "agi", "artificial intelligence", "ai model",
  "llm", "claude", "gemini", "mistral", "anthropic",
  "apple", "nvidia", "meta ", "google", "microsoft", "amazon", "tesla",
  "spacex", "robotics", "autonomous", "self-driving", "quantum",
  "regulation ai", "ai regulation", "deepmind", "sam altman",
];

// ── Ticker-to-keyword mapping for portfolio correlation ──────────────────────

const TICKER_KEYWORDS: Record<string, string[]> = {
  PETR4: ["petrobras", "petróleo", "oil price", "crude", "opec"],
  VALE3: ["vale", "iron ore", "mining", "minério"],
  ITUB4: ["itaú", "itau", "banking", "brazil bank"],
  BBDC4: ["bradesco", "banking", "brazil bank"],
  BBAS3: ["banco do brasil", "banking", "brazil bank"],
  WEGE3: ["weg", "industrial"],
  RENT3: ["localiza", "rental"],
  ABEV3: ["ambev", "beer", "beverage"],
  B3SA3: ["b3", "stock exchange", "bolsa"],
  ELET3: ["eletrobras", "energy", "electricity"],
  SUZB3: ["suzano", "pulp", "paper"],
  JBSS3: ["jbs", "meat", "beef"],
  MGLU3: ["magazine luiza", "e-commerce"],
  HAPV3: ["hapvida", "health"],
  BBSE3: ["bb seguridade", "insurance"],
  IVVB11: ["s&p 500", "s&p500"],
  BOVA11: ["ibovespa"],
  VOO: ["s&p 500", "s&p500", "vanguard"],
  SPY: ["s&p 500", "s&p500"],
  QQQ: ["nasdaq", "tech"],
  AAPL: ["apple"],
  MSFT: ["microsoft"],
  GOOGL: ["google", "alphabet"],
  AMZN: ["amazon"],
  NVDA: ["nvidia"],
  META: ["meta ", "facebook", "instagram"],
  TSLA: ["tesla", "elon musk"],
  BTC: ["bitcoin", "btc"],
  ETH: ["ethereum"],
  VWRA: ["global equity", "world index"],
  ASML: ["asml", "semiconductor", "chip"],
  DPM: ["diversified royalties"],
};

// ── Helpers ───────────────────────────────────────────────────────────────────

function daysRemaining(endDate: string): number | null {
  if (!endDate) return null;
  try {
    const dt = new Date(endDate);
    const delta = Math.floor((dt.getTime() - Date.now()) / 86400000);
    return delta < 0 ? null : delta;
  } catch {
    return null;
  }
}

const BINARY = new Set(["yes", "no", "sim", "não", "nao"]);

function parseJsonField(raw: unknown): string[] {
  if (Array.isArray(raw)) return raw.map(String);
  if (typeof raw === "string") {
    try { return JSON.parse(raw); } catch { return []; }
  }
  return [];
}

function labelFromMarket(m: Record<string, any>): string {
  const group = (m.groupItemTitle || "").trim();
  if (group) return group;
  const outcomes = parseJsonField(m.outcomes);
  for (const o of outcomes) {
    if (!BINARY.has(o.trim().toLowerCase())) return o.trim();
  }
  let q = (m.question || "").trim().replace(/\?$/, "");
  for (const p of ["Will ", "Does ", "Is ", "Can ", "Has ", "Did ", "Do "]) {
    if (q.startsWith(p)) { q = q.slice(p.length); break; }
  }
  return q.slice(0, 50);
}

function yesPrice(m: Record<string, any>): number | null {
  const outcomes = parseJsonField(m.outcomes);
  const prices = parseJsonField(m.outcomePrices);
  if (!outcomes.length || !prices.length) return null;
  const yesIdx = outcomes.findIndex((o: string) => ["yes", "sim"].includes(o.trim().toLowerCase()));
  const idx = yesIdx >= 0 ? yesIdx : 0;
  try {
    const pv = parseFloat(prices[idx]);
    if (pv > 1) return pv / 100;
    return pv > 0 && pv < 1 ? pv : null;
  } catch {
    return null;
  }
}

function normalizePrice(pv: number): number {
  if (pv > 1 && pv <= 100) return pv / 100;
  return pv;
}

// ── Fetch + parse (runs client-side in the browser) ──────────────────────────

function parseEventList(data: Record<string, any>[]): PolyEvent[] {
  const processed: PolyEvent[] = [];

  for (const event of data) {
    try {
      const markets: Record<string, any>[] = event.markets ?? [];
      if (!markets.length) continue;

      const totalVolume = markets.reduce((s: number, m: any) => {
        const v = typeof m.volume === "number" ? m.volume : parseFloat(m.volume || "0") || 0;
        return s + v;
      }, 0);

      const odds: PolyOdds[] = [];

      if (markets.length === 1) {
        const m = markets[0];
        const outcomes = parseJsonField(m.outcomes);
        const prices = parseJsonField(m.outcomePrices);
        for (let i = 0; i < outcomes.length; i++) {
          let pv = parseFloat(prices[i]);
          if (isNaN(pv)) continue;
          pv = normalizePrice(pv);
          if (pv > 0 && pv < 1) {
            odds.push({ outcome: outcomes[i], price: pv, percent: Math.round(pv * 1000) / 10 });
          }
        }
      } else {
        for (const m of markets) {
          const label = labelFromMarket(m);
          if (!label) continue;
          const pv = yesPrice(m);
          if (pv === null) continue;
          odds.push({ outcome: label, price: pv, percent: Math.round(pv * 1000) / 10 });
        }
      }

      odds.sort((a, b) => b.percent - a.percent);
      if (!odds.length || odds[0].percent >= 99) continue;

      processed.push({
        id: String(event.id ?? ""),
        title: String(event.title ?? "Unknown Event"),
        url: `https://polymarket.com/event/${event.slug ?? ""}`,
        volume: totalVolume,
        end_date: String(event.endDate ?? ""),
        days_left: daysRemaining(String(event.endDate ?? "")),
        odds,
        is_binary: odds.length === 2,
      });
    } catch {
      continue;
    }
  }

  processed.sort((a, b) => b.volume - a.volume);
  return processed;
}

async function fetchEvents(): Promise<{ events: PolyEvent[]; totalFetched: number }> {
  const url = `${GAMMA_API}?limit=200&active=true&closed=false&order=volume_24hr&ascending=false`;
  const res = await fetch(url, {
    headers: { Accept: "application/json" },
    signal: AbortSignal.timeout(15000),
  });
  if (!res.ok) throw new Error(`Polymarket API ${res.status}`);
  let data = await res.json();
  if (!Array.isArray(data)) data = data?.data ?? data?.events ?? [];

  return { events: parseEventList(data), totalFetched: data.length };
}

// ── Preditivos de PREÇO dos ativos da carteira ───────────────────────────────
// Os top-200 por volume geral quase nunca incluem mercados de preço de um
// ativo específico ("MSFT above $500?"). Busca DIRECIONADA por ativo na API de
// busca do Polymarket e filtra títulos que falam de preço/valuation.

const PRICE_RE = /\$|price|hit\s|reach|above|below|close at|all[- ]time high|market cap|valuation|per share/i;

function searchQueryForTicker(t: string): string | null {
  const clean = t.replace(/\.(SA|L|DE|TO|AS|PA|MI|MC|LS)$/i, "").replace(/-USD$/i, "").toUpperCase();
  const mapped = TICKER_KEYWORDS[clean]?.[0];
  if (mapped) return mapped;
  // B3 sem mapeamento (PETR4-like): Polymarket não cobre — não desperdiça busca.
  if (/^[A-Z]{4}\d{1,2}$/.test(clean)) return null;
  return clean.toLowerCase();
}

async function searchAssetPriceEvents(tickers: string[]): Promise<PolyEvent[]> {
  const queries = [...new Set(tickers.map(searchQueryForTicker).filter((q): q is string => Boolean(q)))].slice(0, 10);
  if (!queries.length) return [];

  const results = await Promise.allSettled(queries.map(async (q) => {
    const url = `https://gamma-api.polymarket.com/public-search?q=${encodeURIComponent(q)}&limit_per_type=10&events_status=active`;
    const res = await fetch(url, { headers: { Accept: "application/json" }, signal: AbortSignal.timeout(10000) });
    if (!res.ok) return [] as PolyEvent[];
    const data = await res.json();
    const list = Array.isArray(data?.events) ? data.events : [];
    return parseEventList(list).filter((ev) => PRICE_RE.test(ev.title));
  }));

  const byId = new Map<string, PolyEvent>();
  for (const r of results) {
    if (r.status !== "fulfilled") continue;
    for (const ev of r.value) if (!byId.has(ev.id)) byId.set(ev.id, ev);
  }
  return [...byId.values()].sort((a, b) => b.volume - a.volume);
}

function classify(events: PolyEvent[]): Record<string, PolyEvent[]> {
  const cats: Record<string, PolyEvent[]> = {
    "🏦 Macro & Finanças": [],
    "🌍 Geopolítica": [],
    "🤖 Tech & IA": [],
    "⭐ Em Destaque": [],
  };

  for (const ev of events) {
    const text = (ev.title + " ").toLowerCase();
    if (CRYPTO_KW.some(kw => text.includes(kw))) continue;

    if (MACRO_KW.some(kw => text.includes(kw))) cats["🏦 Macro & Finanças"].push(ev);
    else if (GEO_KW.some(kw => text.includes(kw))) cats["🌍 Geopolítica"].push(ev);
    else if (TECH_AI_KW.some(kw => text.includes(kw))) cats["🤖 Tech & IA"].push(ev);
    else cats["⭐ Em Destaque"].push(ev);
  }

  for (const k of Object.keys(cats)) cats[k] = cats[k].slice(0, 15);
  return Object.fromEntries(Object.entries(cats).filter(([, v]) => v.length > 0));
}

function correlate(events: PolyEvent[], tickers: string[]): PolyEvent[] {
  const keywords: string[] = [];
  for (const t of tickers) {
    const clean = t.replace(".SA", "").replace(".L", "").replace(".DE", "").replace(".TO", "").replace(".AS", "").toUpperCase();
    const mapped = TICKER_KEYWORDS[clean];
    if (mapped) keywords.push(...mapped);
    else keywords.push(clean.toLowerCase());
  }
  if (!keywords.length) return [];

  const matched: PolyEvent[] = [];
  for (const ev of events) {
    const text = (ev.title + " ").toLowerCase();
    if (keywords.some(kw => text.includes(kw))) {
      matched.push(ev);
    }
  }
  return matched.slice(0, 15);
}

// ── Portfolio impact mapping ────────────────────────────────────────────────

const IMPACT_KW: Record<string, string[]> = {
  "fed ": ["VOO", "SPY", "QQQ", "IVVB11"],
  "interest rate": ["ITUB4", "BBDC4", "BBAS3", "B3SA3"],
  "inflation": ["BOVA11", "IVVB11", "VOO"],
  "recession": ["BOVA11", "VOO", "SPY"],
  "oil": ["PETR4"], "crude": ["PETR4"],
  "s&p": ["VOO", "SPY", "IVVB11"],
  "nasdaq": ["QQQ"],
  "bitcoin": ["BTC"], "ethereum": ["ETH"],
  "nvidia": ["NVDA"], "apple": ["AAPL"], "tesla": ["TSLA"],
  "microsoft": ["MSFT"], "google": ["GOOGL"], "amazon": ["AMZN"],
  "china": ["VALE3"], "iron ore": ["VALE3"],
  "tariff": ["VALE3", "SUZB3", "JBSS3", "PETR4"],
  "brazil": ["BOVA11", "IVVB11"],
};

export function findPortfolioImpact(title: string): string[] {
  const text = title.toLowerCase();
  const hits = new Set<string>();
  for (const [kw, tickers] of Object.entries(IMPACT_KW)) {
    if (text.includes(kw)) tickers.forEach(t => hits.add(t));
  }
  return [...hits];
}

// ── Unified prediction event type ──────────────────────────────────────────

export interface UnifiedPrediction {
  id: string;
  source: "polymarket" | "kalshi" | "metaculus";
  title: string;
  url: string;
  category: string;
  odds: { outcome: string; percent: number }[];
  volume?: number;
  forecasters?: number;
  end_date: string;
  days_left: number | null;
  portfolio_impact: string[];
}

export function polyToUnified(ev: PolyEvent): UnifiedPrediction {
  return {
    id: ev.id,
    source: "polymarket",
    title: ev.title,
    url: ev.url,
    category: "",
    odds: ev.odds.map(o => ({ outcome: o.outcome, percent: o.percent })),
    volume: ev.volume,
    end_date: ev.end_date,
    days_left: ev.days_left,
    portfolio_impact: findPortfolioImpact(ev.title),
  };
}

// ── Public API ───────────────────────────────────────────────────────────────

export const PRECO_ATIVOS_CAT = "💲 Preço dos Ativos";

export async function fetchPolymarket(portfolioTickers: string[] = []): Promise<PolyResponse> {
  // Busca direcionada por ativo em paralelo com o feed geral; falha da busca
  // não derruba o resto (cai só a categoria de preço).
  const pricePromise = portfolioTickers.length
    ? searchAssetPriceEvents(portfolioTickers).catch(() => [] as PolyEvent[])
    : Promise.resolve([] as PolyEvent[]);
  const { events, totalFetched } = await fetchEvents();
  const priceEvents = await pricePromise;

  // Cada evento vive em UMA categoria (a página achata as categorias — id
  // duplicado viraria card duplicado).
  const priceIds = new Set(priceEvents.map((e) => e.id));
  const rest = events.filter((e) => !priceIds.has(e.id));

  const categories: Record<string, PolyEvent[]> = {};
  if (priceEvents.length > 0) categories[PRECO_ATIVOS_CAT] = priceEvents.slice(0, 15);
  Object.assign(categories, classify(rest));

  if (portfolioTickers.length > 0) {
    const correlated = correlate(rest, portfolioTickers);
    if (correlated.length > 0) {
      categories["📊 Correlatos ao Portfólio"] = correlated;
    }
  }

  return {
    categories,
    cached_at: new Date().toISOString(),
    total_fetched: totalFetched,
    total_parsed: events.length + priceEvents.length,
  };
}

export { fetchEvents, classify, correlate };
